feat: U15 Sepolia - Base (#873)
ethereum-optimism/superchain-ops: src/improvements/tasks/sep/011-opcm-upgrade-prestate-v300-base/VALIDATION.md

# Validation

This document can be used to validate the inputs and result of the execution of the upgrade transaction which you are
signing.

The steps are:

1. [Validate the Domain and Message Hashes](#expected-domain-and-message-hashes)
2. [Verifying the transaction input](#understanding-task-calldata)
3. [Verifying the state changes](#state-changes)

## Expected Domain and Message Hashes

First, we need to validate the domain and message hashes. These values should match both the values on your ledger and
the values printed to the terminal when you run the task.

> [!CAUTION]
>
> Before signing, ensure the below hashes match what is on your ledger.
>
> ### Nested Safe 1: `0x6AF0674791925f767060Dd52f7fB20984E8639d8`
>

> > - Domain Hash: `0x6f25427e79742a1eb82c103e2bf43c85fc59509274ec258ad6ed841c4a0048aa`

Reviewer: I get this in my terminal:

```
------------------ Nested Multisig EOAs Hash to Approve ------------------
  Child multisig: SUPERCHAINADDRESSREGISTRY_ChildSafe1 @0x6AF0674791925f767060Dd52f7fB20984E8639d8
  Domain Hash:     0x6f25427e79742a1eb82c103e2bf43c85fc59509274ec258ad6ed841c4a0048aa
  Message Hash:    0xff5166e1a86ea2e7a78192844ea6332b1357d136f1de36bd4230a70fc952d1c4
  ```

> - Message Hash: `0x4cdac20368824e0bf4725fa51adbab279fb9e378750085a2255937b69e4b7c06`
>
> ### Nested Safe 2: `0x646132A1667ca7aD00d36616AFBA1A28116C770A`
>

> > - Domain Hash: `0x1d3f2566fd7b1bf017258b03d4d4d435d326d9cb051d5b7993d7c65e7ec78d0e`

Reviewer: I get this:
```
------------------ Nested Multisig EOAs Hash to Approve ------------------
  Child multisig: SUPERCHAINADDRESSREGISTRY_ChildSafe2 @0x646132A1667ca7aD00d36616AFBA1A28116C770A
  Domain Hash:     0x1d3f2566fd7b1bf017258b03d4d4d435d326d9cb051d5b7993d7c65e7ec78d0e
  Message Hash:    0xff5166e1a86ea2e7a78192844ea6332b1357d136f1de36bd4230a70fc952d1c4
  ```

> - Message Hash: `0x4cdac20368824e0bf4725fa51adbab279fb9e378750085a2255937b69e4b7c06`

## Understanding Task Calldata

This document provides a detailed analysis of the final calldata executed on-chain for the OPCM upgrade to v2.0.0.

By reconstructing the calldata, we can confirm that the execution precisely implements the approved upgrade plan with no unexpected modifications or side effects.

### Inputs to `opcm.upgrade()`

> For each chain being upgrade, the `opcm.upgrade()` function is called with a tuple of three elements:

Reviewer: There are still references to the upgrade function in this file. In the OP,Ink VALIDATION file I renamed all of these.

> For each chain being upgrade, the `opcm.upgrade()` function is called with a tuple of three elements:

Reviewer: ```suggestion
For each chain being upgrade, the `opcm.updatePrestate()` function is called with a tuple of a single element:
```

1. Base Sepolia Testnet:
   - SystemConfigProxy: [0xf272670eb55e895584501d564AfEB048bEd26194](https://github.com/ethereum-optimism/superchain-registry/blob/2c60e5723c64b5a1b58ab72c5d3816927ff9391a/superchain/configs/sepolia/base.toml#L59)
   - ProxyAdmin: [0x0389E59Aa0a41E4A413Ae70f0008e76CAA34b1F3](https://github.com/ethereum-optimism/superchain-registry/blob/2c60e5723c64b5a1b58ab72c5d3816927ff9391a/superchain/configs/sepolia/base.toml#L60)

>    - AbsolutePrestate: [0x03682932cec7ce0a3874b19675a6bbc923054a7b321efc7d3835187b172494b6](https://github.com/ethereum-optimism/superchain-registry/blob/712a84f44501322ca61901c2729aa3a56726a602/validation/standard/standard-prestates.toml#L10)

Reviewer: This prestate is not correct.

Author: Why not? 
It matches what I [see here](https://www.notion.so/oplabs/Isthmus-Sepolia-Mainnet-1d2f153ee162800880abe1b47910c071?pvs=4).

Reviewer: Ah, my bad I thought the link was incorrect. Resolving this comment.

```

### Inputs to `Multicall3DelegateCall`

The output from the previous section becomes the `data` in the argument to the `Multicall3DelegateCall.aggregate3()` function.

This function is called with a tuple of three elements:

Call3 struct for Multicall3DelegateCall:

- `target`: [0xfBceeD4DE885645fBdED164910E10F52fEBFAB35](https://github.com/ethereum-optimism/superchain-registry/blob/main/validation/standard/standard-versions-sepolia.toml#L22) - Sepolia OPContractsManager v3.0.0
- `allowFailure`: false
- `callData`: `0xff2dd5a1...` (output from the previous section)

Command to encode:

```bash
cast calldata 'aggregate3((address,bool,bytes)[])' "[(0xfBceeD4DE885645fBdED164910E10F52fEBFAB35,false,0x9a72745b00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000f272670eb55e895584501d564afeb048bed261940000000000000000000000000389e59aa0a41e4a413ae70f0008e76caa34b1f303682932cec7ce0a3874b19675a6bbc923054a7b321efc7d3835187b172494b6)]"
```

The resulting calldata sent from the ProxyAdminOwner safe is thus:

```

> 0x82ad56cb000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000fbceed4de885645fbded164910e10f52febfab350000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a49a72745b00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000f272670eb55e895584501d564afeb048bed261940000000000000000000000000389e59aa0a41e4a413ae70f0008e76caa34b1f303682932cec7ce0a3874b19675a6bbc923054a7b321efc7d3835187b172494b600000000000000000000000000000000000000000000000000000000

Reviewer: I'm not getting this call data in the terminal: 

I get:
```
0x82ad56cb000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000020000000000000000000000000fbceed4de885645fbded164910e10f52febfab350000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a49a72745b00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000001000000000000000000000000f272670eb55e895584501d564afeb048bed261940000000000000000000000000389e59aa0a41e4a413ae70f0008e76caa34b1f303682932cec7ce0a3874b19675a6bbc923054a7b321efc7d3835187b172494b600000000000000000000000000000000000000000000000000000000
```

Reviewer: OK I'm closing this as it's not matching.

```

> In mainnet runbooks, this calldata should appear in [Action Plan](https://gov.optimism.io/t/upgrade-proposal-14-isthmus-l1-contracts-mt-cannon/9796#p-43948-action-plan-9) section of the Governance proposal.

Reviewer: This points to the U14 gov post.

  - **Before:** `0x0000000000000000000000000000000000000000000000000000000000000000`
  - **After:** `0x0000000000000000000000000000000000000000000000000000000000000001`
  - **Summary:**  `approveHash(bytes32)` called on ProxyAdminOwner by child multisig.
  - **Detail:** **THIS WAS CAREFULLY VERIFIED BY RUNBOOK REVIEWERS AND NEED NOT BE CHECKED BY SIGNERS.** As part of the Tenderly simulation, we want to illustrate the <i>approveHash</i> invocation. This step isn't shown in the local simulation because the parent multisig is invoked directly, bypassing the <i>approveHash</i> calls. This slot change reflects an update to the approvedHashes mapping.
    Specifically, this simulation was ran as the nested safe ``. To verify the slot yourself, run:
    - `res=$(cast index address 0x646132A1667ca7aD00d36616AFBA1A28116C770A 8)`
    - `cast index bytes32 0xf7f42d9617b2d58b07ebbba60ac231142f29f643c82bef5752250bcaa0ef9b34 $res`
    - Please note: the `0xf7f42d9617b2d58b07ebbba60ac231142f29f643c82bef5752250bcaa0ef9b34` value is taken from the Tenderly simulation and this is the transaction hash of the `approveHash` call.

>   ---

Reviewer: I see this state diff in here:
![CleanShot 2025-04-11 at 17 45 38](https://github.com/user-attachments/assets/2a602cf5-52a5-47f2-bfd0-1deb71b13c30)

Reviewer: And when signing for child-safe-2 I see this one:
![CleanShot 2025-04-11 at 17 46 26](https://github.com/user-attachments/assets/ae35a52b-7537-42a2-9370-ca5161487388)

### [`0xd6e6dbf4f7ea0ac412fd8b65ed297e64bb7a06e1`](https://github.com/ethereum-optimism/superchain-registry/blob/08e3fe429c776a532c2b6dc09571fc13e6dba5d4/superchain/configs/sepolia/base.toml#L63)  (DisputeGameFactory) - Chain ID: 84532

- **Key:**          `0x4d5a9bd2e41301728d41c8e705190becb4e74abe869f75bdb405b63716a35f9e`
  - **Before:**     `0x000000000000000000000000d53394d4f67653074acf0b264954fe5e4f72d24f`
  - **After:**     `0x000000000000000000000000b63663de7b5f70d50ed841dc04e62e3bbd64e72d`
  - **Summary:** Updates the implementation for game type 1.

>   - **Detail:** This is `gameImpls[1]` -> `0x97766954BAF17e3a2BfA43728830f0Fa647F7546`. The [`gameImpls` mapping](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol#L57) is at [storage slot 101](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/snapshots/storageLayout/DisputeGameFactory.json#L41) and is keyed by [`GameType` (`uint32`)](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/lib/LibUDT.sol#L224).

Reviewer: ```suggestion
  - **Detail:** This is `gameImpls[1]` -> `0xb63663de7b5f70d50ed841dc04e62e3bbd64e72d`. The [`gameImpls` mapping](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol#L57) is at [storage slot 101](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/snapshots/storageLayout/DisputeGameFactory.json#L41) and is keyed by [`GameType` (`uint32`)](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/lib/LibUDT.sol#L224).
```

    - Confirm the expected key slot with the following:
      ```shell
      cast index uint32 1 101
      0x4d5a9bd2e41301728d41c8e705190becb4e74abe869f75bdb405b63716a35f9e
      ```

- **Key:**          `0xffdfc1249c027f9191656349feb0761381bb32c9f557e01f419fd08754bf5a1b`
  - **Before:**     `0x000000000000000000000000861eb6dfe0fde8c8a63e8606fa487ee870f65e72`
  - **After:**     `0x00000000000000000000000097ca5600adc6a198df4c82ff9dccb74384ed3641`

>   - **Detail:** This is `gameImpls[0]` -> `0xBd72dD2fB74a537B9B47B454614A15B066Cc464a`. The [`gameImpls` mapping](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol#L57) is at [storage slot 101](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/snapshots/storageLayout/DisputeGameFactory.json#L41) and is keyed by [`GameType` (`uint32`)](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/lib/LibUDT.sol#L224).

Reviewer: ```suggestion
  - **Detail:** This is `gameImpls[0]` -> `0x97ca5600adc6a198df4c82ff9dccb74384ed3641`. The [`gameImpls` mapping](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/DisputeGameFactory.sol#L57) is at [storage slot 101](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/snapshots/storageLayout/DisputeGameFactory.json#L41) and is keyed by [`GameType` (`uint32`)](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v3.0.0-rc.2/packages/contracts-bedrock/src/dispute/lib/LibUDT.sol#L224).
```
